examples/coap-server.c

```diff
@@ -377,9 +380,196 @@ check_async(coap_context_t *ctx,
+        if ((i % 10) == 0) {
+          example_data_ptr->s[i] = 'a' + (i/10) % 26;
+        }
+        else {
+          example_data_ptr->s[i] = '0' + i%10;
+        }
+      }
+    }
+  }
+  coap_add_data_blocked_response(resource, session, request, response, token,
+                                 example_data_media_type, -1,
+                                 example_data_ptr ? example_data_ptr->length : 0,
+                                 example_data_ptr ? example_data_ptr->s : NULL);
+  return;
+}
+
+static void
+cache_free_app_data(void *data) {
+  coap_binary_t *bdata = (coap_binary_t*)data;
+  coap_delete_binary(bdata);
+}
+
+/*
+ * Large Data PUT handler
+ */
+
+static void
+hnd_put_example_data(coap_context_t *ctx UNUSED_PARAM,
+        coap_resource_t *resource,
+        coap_session_t *session,
+        coap_pdu_t *request,
+        coap_binary_t *token UNUSED_PARAM,
+        coap_string_t *query UNUSED_PARAM,
+        coap_pdu_t *response
+) {
+  size_t size;
+  uint8_t *data;
+  coap_block_t block1;
+  unsigned char buf[6];      /* space to hold encoded/decoded uints */
+  coap_opt_iterator_t opt_iter;
+  coap_opt_t *option;
+
+  if (coap_get_block(request, COAP_OPTION_BLOCK1, &block1)) {
+    /* handle BLOCK1 */
+    coap_cache_entry_t *cache_entry = coap_cache_get_by_pdu(session,
+                                                            request,
+                                                COAP_CACHE_IS_SESSION_BASED);
+    size_t offset = block1.num << (block1.szx + 4);
+    coap_binary_t *data_so_far;
+
+    if (!cache_entry && block1.num == 0) {
+      cache_entry = coap_new_cache_entry(session, request, 0, 1, -1);
+    }
+    if (!cache_entry) {
+      if (block1.num == 0) {
+        coap_log(LOG_WARNING, "Unable to create a new cache entry\n");
+      }
+      else {
+        coap_log(LOG_WARNING,
+                 "No cache entry available for the non-first BLOCK1\n");
+      }
+      response->code = COAP_RESPONSE_CODE(500);
+      return;
+    }
+
+    data_so_far = coap_cache_get_app_data(cache_entry);
+    if (offset == 0) {
+      if (data_so_far) {
+        coap_delete_binary(data_so_far);
+        data_so_far = NULL;
+      }
+    }
+    else if (offset >
+             (data_so_far ? data_so_far->length : 0)) {
+      /* Upload is not sequential - block missing */
+      response->code = COAP_RESPONSE_CODE(408);
+      return;
+    }
+    else if (offset <
+             (data_so_far ? data_so_far->length : 0)) {
+      /* Upload is not sequential - block duplicated */
+      goto just_respond;
+    }
+
+    if (coap_get_data(request, &size, &data) && (size > 0)) {
+      if (!data_so_far) {
+        data_so_far = coap_new_binary(size);
+        memcpy(data_so_far->s, data, size);
+      }
+      else {
+        /* Add in new block to end of current data */
+        data_so_far = coap_resize_binary(data_so_far, offset + size);
+        memcpy(&data_so_far->s[offset], data, size);
```

Comment:
Here, `coap_resize_binary()` may return `NULL` so that `data_so_far` must be checked before `memcpy()`.

Reply:
All calls to coap_resize_binary() now checked in coap-server.